hedera-node/src/main/java/com/hedera/services/txns/token/TokenUpdateTransitionLogic.java

```diff
@@ -65,125 +67,195 @@
 	private final Function<TransactionBody, ResponseCodeEnum> SEMANTIC_CHECK = this::validate;
 
 	private final boolean allowChangedTreasuryToOwnNfts;
-	private final TokenStore store;
+	private final TransactionContext txnCtx;
 	private final HederaLedger ledger;
 	private final OptionValidator validator;
-	private final TransactionContext txnCtx;
 	private final Predicate<TokenUpdateTransactionBody> affectsExpiryOnly;
+	private final TypedTokenStore tokenStore;
+	private final AccountStore accountStore;
 
 	public TokenUpdateTransitionLogic(
 			boolean allowChangedTreasuryToOwnNfts,
-			OptionValidator validator,
-			TokenStore store,
 			HederaLedger ledger,
+			OptionValidator validator,
 			TransactionContext txnCtx,
-			Predicate<TokenUpdateTransactionBody> affectsExpiryOnly
+			Predicate<TokenUpdateTransactionBody> affectsExpiryOnly,
+			TypedTokenStore tokenStore,
+			AccountStore accountStore
 	) {
 		this.validator = validator;
-		this.store = store;
-		this.ledger = ledger;
 		this.txnCtx = txnCtx;
 		this.affectsExpiryOnly = affectsExpiryOnly;
 		this.allowChangedTreasuryToOwnNfts = allowChangedTreasuryToOwnNfts;
+		this.ledger = ledger;
+		this.tokenStore = tokenStore;
+		this.accountStore = accountStore;
 	}
 
 	@Override
 	public void doStateTransition() {
-		try {
-			transitionFor(txnCtx.accessor().getTxn().getTokenUpdate());
-		} catch (Exception e) {
-			log.warn("Unhandled error while processing :: {}!", txnCtx.accessor().getSignedTxnWrapper(), e);
-			abortWith(FAIL_INVALID);
+		final var op = txnCtx.accessor().getTxn().getTokenUpdate();
+		final var token = tokenStore.loadToken(Id.fromGrpcToken(op.getToken()));
+
+		throwIfExpirationIsInvalid(op);
+		throwIfTokenIsImmutable(op, token);
+		throwIfCurrentAutoRenewAccountIsDetached(token);
+
+		final var optionalNewAutoRenewAccount =
+				Optional.ofNullable(loadNewAutorenewAccountOrThrowIfInvalid(op));
+
+		final var optionalNewTreasury =
+				Optional.ofNullable(loadNewTreasuryOrThrowIfInvalid(op));
+
+		Account newTreasury = null;
+		TokenRelationship newTreasuryRel = null;
+
+		if (optionalNewTreasury.isPresent()) {
+			final var currentTreasury = accountStore.loadAccount(token.getTreasury().getId());
+			newTreasury = optionalNewTreasury.get();
+
+			throwIfAssociationIsMissingBetween(newTreasury, token);
+			throwIfCurrentTreasuryOwnsNfts(token);
+
+			newTreasuryRel = tokenStore.loadTokenRelationship(token, newTreasury);
+
+			final var treasuriesAreNotEqual = !currentTreasury.getId().equals(newTreasury.getId());
+
+			if (treasuriesAreNotEqual) {
+				prepareNewTreasury(newTreasuryRel, token);
```

Comment:
Instead of having `prepareNewTreasury` method that changes `TokenRel` properties, I think that it would be better to have `newTreasuryRel.prepareForTreasury` method or use the already existing method. I think that `TokenCreate` uses a method that can be reused in this case